pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeSegmentPruner.java

```diff
@@ -0,0 +1,344 @@
+    int i = 0;
+    int j = 0;
+    while (i < size1 && j < size2) {
+      Interval interval1 = intervals1.get(i);
+      Interval interval2 = intervals2.get(j);
+      if (interval1.intersects(interval2)) {
+        res.add(Interval.getIntersection(interval1, interval2));
+      }
+      if (interval1._max <= interval2._max) {
+        i++;
+      } else {
+        j++;
+      }
+    }
+    return res;
+  }
+
+  private List<Interval> getUnionSortedIntervals(List<List<Interval>> intervals) {
+    // Requires input intervals are sorted, the return intervals will be sorted
+    if (intervals.isEmpty()) {
+      return null;
+    }
+    return getUnionSortedIntervals(intervals, 0, intervals.size());
+  }
+
+  private List<Interval> getUnionSortedIntervals(List<List<Interval>> intervals, int start, int end) {
+    if (start + 1 == end) {
+      return intervals.get(start);
+    }
+
+    int mid = start + (end - start) / 2;
+    List<Interval> intervals1 = getUnionSortedIntervals(intervals, start, mid);
+    List<Interval> intervals2 = getUnionSortedIntervals(intervals, mid, end);
+    return getUnionTwoSortedIntervals(intervals1, intervals2);
+  }
+
+  /**
+   * Union two list of non-overlapping sorted intervals.
+   * E.g. {[1, 2], [5, 7], [9, 10]} and {[2, 3], [4, 8]} are merged as {[1, 3], [4, 8], [9, 10]}
+   */
+  private List<Interval> getUnionTwoSortedIntervals(List<Interval> intervals1, List<Interval> intervals2) {
+    List<Interval> res = new ArrayList<>();
+    int size1 = intervals1.size();
+    int size2 = intervals2.size();
+    int i = 0;
+    int j = 0;
+    while (i < size1 || j < size2) {
+      // Get the `smaller` interval
+      Interval interval = null;
+      if (j == size2 || i < size1 && intervals1.get(i).compareTo(intervals2.get(j)) <= 0) {
+        interval = intervals1.get(i++);
+      } else {
+        interval = intervals2.get(j++);
+      }
+      // If not overlapping with current result, add as a new interval
+      int resSize = res.size();
+      if (res.isEmpty() || !interval.intersects(res.get(resSize - 1))) {
+        res.add(interval);
+      } else {
+        // If overlaps with the result, union with current result
+        res.set(resSize - 1, Interval.getUnion(interval, res.get(resSize - 1)));
+      }
+    }
+    return res;
+  }
+
+  /**
+   * Parse interval to millisecond as [min, max] with both sides included.
+   * E.g. '(* 16311]' is parsed as [0, 16311], '(1455 16311)' is parsed as [1456, 16310]
+   */
+  private List<Interval> parseInterval(List<String> intervalExpressions) {
+    Preconditions.checkState(intervalExpressions != null && intervalExpressions.size() == 1,
+        "Cannot parse range expressions from query: %s", intervalExpressions);
+    long startTime = MIN_START_TIME;
+    long endTime = MAX_END_TIME;
+    String intervalExpression = intervalExpressions.get(0);
+    String[] s = intervalExpression.substring(1, intervalExpression.length() - 1).split("\\s+");
```

Comment:
(Critical) Don't use arbitrary whitespace as the delimiter (they might be valid as the value, also I don't think '\0` is a whitespace). You may refer to `RangePredicate.java` on how to parse the range string

Reply:
Fixed. Thank you for helping to review and improve the code, appreciate it!